File: ios/RocChat/Crypto/KeyRotationManager.swift
import CryptoKit
import Foundation

/// Handles SPK rotation and one-time prekey replenishment on iOS.
/// Runs on app launch / foreground — non-critical, best-effort.
class KeyRotationManager {
    static let shared = KeyRotationManager()
    private let api = APIClient.shared

    /// SPK rotation interval: 7 days
    private let spkRotationInterval: TimeInterval = 7 * 24 * 3600
    /// Minimum OTP keys before replenishment
    private let preKeyThreshold = 10
    /// Number of keys to upload when replenishing
    private let preKeyBatchSize = 20

    // MARK: - SPK Rotation

    func maybeRotateSignedPreKey() async {
        let lastRotation = UserDefaults.standard.double(forKey: "rocchat_spk_last_rotation")
        guard Date().timeIntervalSince1970 - lastRotation >= spkRotationInterval else { return }

        guard let identityPrivB64 = SecureStorage.shared.get(forKey: "identity_priv"),
              let identityPrivData = Data(base64Encoded: identityPrivB64) else { return }

        do {
            // Generate new X25519 keypair for SPK
            let spkPriv = Curve25519.KeyAgreement.PrivateKey()
            let spkPub = spkPriv.publicKey.rawRepresentation

            // Sign with Ed25519 identity key
            let signingKey = try Curve25519.Signing.PrivateKey(rawRepresentation: identityPrivData)
            let signature = try signingKey.signature(for: spkPub)
            let spkId = Int(Date().timeIntervalSince1970)

            let body: [String: Any] = [
                "id": spkId,
                "publicKey": spkPub.base64EncodedString(),
                "signature": signature.base64EncodedString(),
            ]
            _ = try await api.postRaw("/keys/signed", body: body, method: "PUT")

            UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: "rocchat_spk_last_rotation")
            SecureStorage.shared.set(spkPriv.rawRepresentation.base64EncodedString(), forKey: "spk_priv")
        } catch {
            // Non-critical — will retry next launch
        }
    }

    // MARK: - Prekey Replenishment

    func checkAndReplenishPreKeys() async {
        do {
            let data = try await api.getRaw("/keys/prekey-count")
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let count = json["count"] as? Int else { return }

            guard count < preKeyThreshold else { return }

            var keys: [[String: Any]] = []
            let baseId = Int(Date().timeIntervalSince1970)
            for i in 0..<preKeyBatchSize {
                let kp = Curve25519.KeyAgreement.PrivateKey()
                keys.append([
                    "id": baseId + i,
                    "publicKey": kp.publicKey.rawRepresentation.base64EncodedString(),
                ])
            }

            _ = try await api.postRaw("/keys/prekeys", body: ["keys": keys])
        } catch {
            // Non-critical
        }
    }

    /// Call on app launch / foreground
    func performMaintenance() async {
        await maybeRotateSignedPreKey()
        await checkAndReplenishPreKeys()
    }
}

File: android/app/src/main/kotlin/com/rocchat/crypto/KeyRotationManager.kt
package com.rocchat.crypto

import android.content.Context
import com.rocchat.network.APIClient
import org.json.JSONArray
import org.json.JSONObject
import java.security.KeyPairGenerator
import java.security.spec.ECGenParameterSpec

/**
 * Handles SPK rotation and one-time prekey replenishment on Android.
 * Runs on app launch — non-critical, best-effort.
 */
object KeyRotationManager {
    private const val SPK_ROTATION_INTERVAL_MS = 7L * 24 * 3600 * 1000 // 7 days
    private const val PRE_KEY_THRESHOLD = 10
    private const val PRE_KEY_BATCH_SIZE = 20

    suspend fun performMaintenance(context: Context) {
        maybeRotateSignedPreKey(context)
        checkAndReplenishPreKeys(context)
    }

    private suspend fun maybeRotateSignedPreKey(context: Context) {
        val prefs = context.getSharedPreferences("rocchat_keys", Context.MODE_PRIVATE)
        val lastRotation = prefs.getLong("spk_last_rotation", 0)
        if (System.currentTimeMillis() - lastRotation < SPK_ROTATION_INTERVAL_MS) return

        val identityPrivB64 = prefs.getString("identity_priv", null) ?: return

        try {
            // Generate new X25519 keypair
            val kpg = KeyPairGenerator.getInstance("XDH")
            kpg.initialize(ECGenParameterSpec("X25519"))
            val kp = kpg.generateKeyPair()
            val pubBytes = kp.public.encoded // X.509 encoded
            // Extract raw 32 bytes from X.509 encoding (last 32 bytes)
            val rawPub = pubBytes.takeLast(32).toByteArray()
            val pubB64 = android.util.Base64.encodeToString(rawPub, android.util.Base64.NO_WRAP)

            // Sign with identity key
            val identityPrivBytes = android.util.Base64.decode(identityPrivB64, android.util.Base64.NO_WRAP)
            val signature = RocCrypto.sign(identityPrivBytes, rawPub)
            val sigB64 = android.util.Base64.encodeToString(signature, android.util.Base64.NO_WRAP)

            val spkId = (System.currentTimeMillis() / 1000).toInt()
            val body = JSONObject().apply {
                put("id", spkId)
                put("publicKey", pubB64)
                put("signature", sigB64)
            }

            APIClient.post("/keys/signed", body, method = "PUT")

            prefs.edit()
                .putLong("spk_last_rotation", System.currentTimeMillis())
                .putString("spk_priv", android.util.Base64.encodeToString(kp.private.encoded, android.util.Base64.NO_WRAP))
                .apply()
        } catch (_: Exception) {
            // Non-critical
        }
    }

    private suspend fun checkAndReplenishPreKeys(context: Context) {
        try {
            val resp = APIClient.get("/keys/prekey-count")
            val count = resp.optInt("count", PRE_KEY_THRESHOLD)
            if (count >= PRE_KEY_THRESHOLD) return

            val keys = JSONArray()
            val baseId = (System.currentTimeMillis() / 1000).toInt()
            val kpg = KeyPairGenerator.getInstance("XDH")
            kpg.initialize(ECGenParameterSpec("X25519"))

            for (i in 0 until PRE_KEY_BATCH_SIZE) {
                val kp = kpg.generateKeyPair()
                val rawPub = kp.public.encoded.takeLast(32).toByteArray()
                keys.put(JSONObject().apply {
                    put("id", baseId + i)
                    put("publicKey", android.util.Base64.encodeToString(rawPub, android.util.Base64.NO_WRAP))
                })
            }

            APIClient.post("/keys/prekeys", JSONObject().put("keys", keys))
        } catch (_: Exception) {
            // Non-critical
        }
    }
}
